<!doctype html>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<link rel="icon" type="image/x-icon" href="../../../static/favicon.ico?h=e38ef9dd">
<link rel="stylesheet" href="../../../static/lib/normalize.css?h=ece571f3">
<link rel="stylesheet" href="../../../static/lib/skeleton.css?h=1fd77917">
<link rel="stylesheet" href="../../../static/style.css?h=5db9e889">
<link rel="stylesheet" href="../../../static/lib/medium-editor/medium-editor.min.css?h=17e98ac9">
<link rel="stylesheet" href="../../../static/fonts/anonymous-pro/anonymous-pro.css?h=7b3d4039">
<link rel="stylesheet" href="../../../static/fonts/fa/css/font-awesome.min.css?h=512c7d79">
<link rel="stylesheet" href="../../../static/fonts/ionicons/css/ionicons.min.css?h=f0459300">
<link rel="stylesheet" href="../../../static/fonts/source-sans-pro/source-sans-pro.css?h=eef2e813">

<!-- <script src="../../../static/lib/clipboard.min.js?h=64efc85d"></script> -->
<script src="../../../static/lib/jquery-2.2.1.min.js?h=7cb05e3d"></script>
<script src="../../../static/lib/marked.js?h=acaa528b"></script>
<script src="../../../static/lib/MooTools-Core-1.6.0.js?h=f829fe22"></script>
<script src="../../../static/lib/sprintf.js?h=f8d59971"></script>
<script src="../../../static/lib/medium-editor/medium-editor.min.js?h=d2c3e1bb"></script>
<script src="../../../static/lib/me-markdown.standalone.min.js?h=f8bd91a7"></script>
<script src="../../../static/lr-notebook.js?h=9fb07e66"></script>


<title></title>
<body>
  <header>
    <h1><i class="fa fa-sticky-note-o"></i> nb-Gabe_Fischbach</h1>
    <nav>
      <ul class="nav navbar-nav">
        <li><a  href="../../../">Welcome</a></li>
        
          <li><a  class="active" href="../../../projects/">Projects</a></li>
        
          <li><a  href="../../../calendar/">Calendar</a></li>
        
          <li><a  href="../../../meetings/">Meetings</a></li>
        
          <li><a  href="../../../about/">About</a></li>
        
          <li><a  href="../../../shared/">Shared</a></li>
        
      </ul>
    </nav>
    
  </header>

  <iframe width="0" height="0" border="0" name="frame" style="border-style: none;"></iframe>

  <div class="page">
    
<h2 style="display: inline;"><a class="orange" href="../../../projects/S_epi/">S_epi/</a></h2><br>
<form target="frame" action="/admin/api/browsefs" method="post">
  <input type="hidden" name="path" value="/projects/S_epi/2017-05-29">
  <input type="image" src=../../../static/images/link.png>
</form>
<h2 class="lightblue" style="display: inline;">Phylogenetic Regression</h2> <span style="float:right;" class="midgray">2017-05-29</span> 
<hr>


<h5 style="float:left;">Attachments:</h5>


<div class="row">
  <div class="one column"></div>
  <div class="two columns"><!-- Attachments: --></div>
<div class="nine columns">
    
      
        <a class="green" href="../../../projects/S_epi/2017-05-29/CD8B_IL17A_counts_tree.svg">CD8B_IL17A_counts_tree.svg</a>
        
  
  
  
  
  
  
  
  
  
  
  
  
  
  <i aria-hidden="true" class="fa fa-object-ungroup"></i>
  
  
 
 
  
  
  

      
    
      
        <a class="green" href="../../../projects/S_epi/2017-05-29/DEDD_Tnp_IS110.svg">DEDD_Tnp_IS110.svg</a>
        
  
  
  
  
  
  
  
  
  
  
  
  
  
  <i aria-hidden="true" class="fa fa-object-ungroup"></i>
  
  
 
 
  
  
  

      
    
      
        <a class="green" href="../../../projects/S_epi/2017-05-29/FA_desaturase.svg">FA_desaturase.svg</a>
        
  
  
  
  
  
  
  
  
  
  
  
  
  
  <i aria-hidden="true" class="fa fa-object-ungroup"></i>
  
  
 
 
  
  
  

      
    
      
        <a class="green" href="../../../projects/S_epi/2017-05-29/mean_cg.svg">mean_cg.svg</a>
        
  
  
  
  
  
  
  
  
  
  
  
  
  
  <i aria-hidden="true" class="fa fa-object-ungroup"></i>
  
  
 
 
  
  
  

      
    
      
        <a href="../../../projects/S_epi/2017-05-29/scrapped_data.png"><img id="scrapped_data.png" draggable="true" ondragstart="drag(event)" ondragend="dragend(event)" src="../../../projects/S_epi/2017-05-29/scrapped_data@50x50.png"></a>
      
    
      
        <a class="green" href="../../../projects/S_epi/2017-05-29/sd_cg.svg">sd_cg.svg</a>
        
  
  
  
  
  
  
  
  
  
  
  
  
  
  <i aria-hidden="true" class="fa fa-object-ungroup"></i>
  
  
 
 
  
  
  

      
    
      
        <a class="green" href="../../../projects/S_epi/2017-05-29/TetR_C_8.svg">TetR_C_8.svg</a>
        
  
  
  
  
  
  
  
  
  
  
  
  
  
  <i aria-hidden="true" class="fa fa-object-ungroup"></i>
  
  
 
 
  
  
  

      
    
    <br/>
    
        <a class="green" href="../../../projects/S_epi/2017-05-29/05_29_17.RData">05_29_17.RData</a>
        
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
 
 
  
  
  
,
    
        <a class="green" href="../../../projects/S_epi/2017-05-29/05_29_17.Rhistory">05_29_17.Rhistory</a>
        
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
 
 
  
  
  
,
    
        <a class="green" href="../../../projects/S_epi/2017-05-29/interesting_pfams.csv">interesting_pfams.csv</a>
        
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
 
 
  
  
  
,
    
        <a class="green" href="../../../projects/S_epi/2017-05-29/pfam_residuals.csv">pfam_residuals.csv</a>
        
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
 
 
  
  
  
,
    
        <a class="green" href="../../../projects/S_epi/2017-05-29/pvals_intercept.csv">pvals_intercept.csv</a>
        
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
 
 
  
  
  
,
    
        <a class="green" href="../../../projects/S_epi/2017-05-29/pvals_mean.csv">pvals_mean.csv</a>
        
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
 
 
  
  
  
,
    
        <a class="green" href="../../../projects/S_epi/2017-05-29/pvals_sorted_1.csv">pvals_sorted_1.csv</a>
        
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
 
 
  
  
  
,
    
        <a class="green" href="../../../projects/S_epi/2017-05-29/pvals_sorted_2.csv">pvals_sorted_2.csv</a>
        
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
  
 
 
  
  
  

    
  </div>
</div>


<hr>





<div id="entry">
    <h2>S_epi 5-29-17</h2>
<p>Continuing from yesterday, when I called barplot.phylo4d on the aggregated tree with count data, it produced a barplot. But when I called phyloSignal on the tree to get summary statistics, it gave me an error saying that some inverse matrix didn't exist.</p>
<p>I'm not sure where this is coming from, but the only thing I could really guess was that maybe it was coming from the existence of a few negative counts data points. Because of the data scraping there were two points scraped from the graph that have negative counts for CD8B-IL17A positive cells. This is not possible. So I went back in and changed the two negative counts to be 1e-12. I've re-done what I did yesterday, saving and loading csv's and adding nodes to the summary csv. For reference, once I have the node data, here's the command I've used to sort by node:</p>
<pre><code>df_sorted = df[order(df[,1]), ]
</code></pre>
<p>where node is the first column of df. Now to merge with data_tree. First, grabbing the last two columns of df_sorted:</p>
<pre><code>df_pruned = df_sorted[ , 3:4]
</code></pre>
<p>So when I've tried merging into a final data tree with addData, it's not working. The indices in the sorted version are still rearranged and it's adding to the rows by index...</p>
<p>So to reset the indices in df_sorted, I've done:</p>
<pre><code>rownames(df_sorted) &lt;- NULL
</code></pre>
<p>then re-making df_pruned and calling</p>
<pre><code>tree_final = addData(data_tree, df_pruned)
</code></pre>
<p>and that looks like its worked. Now I've re-called phyloSignal...and I get the same error.</p>
<pre><code>error: inv(): matrix seems singular 
Error in FUN(newX[, i], ...) : inv(): matrix seems singular
</code></pre>
<p>I've called the phyloCorrelogram functions on mean and sd:</p>
<pre><code>mean.cg &lt;- phyloCorrelogram(tree_final, trait = "mean")
sd.cg &lt;- phyloCorrelogram(tree_final, trait = 'sd')
</code></pre>
<p>and plotted the results:</p>
<pre><code>plot(mean.cg)
plot(sd.cg)
</code></pre>
<p>both are not flat which is a good sign. But how can I get back to the actual genome data from here?</p>
<h2>Orthomcl on S. Epi Data</h2>
<p>I think a decent first-pass here could be to re-run the orthomcl pipeline on the S. Epi data.</p>
<p>I've copied over the shell scripts I wrote from HSV and have made some modifications. I've also copied over the genbank_to_fasta python script and made a change so that when there's a missing protein qualifier or annotation, it adds the corresponding line number (protein number?) to the output fasta file. This is because when I'm running the script to make the output fasta files compliant with orthomcl, the script is crashing on the output fasta files with multiple protein_ids that are identically 'missing_protein_qualifer'.</p>
<p>I've run into an issue with the all-vs-all blast. I wrote shell script 4 that initializes the blastdb and runs an all-vs-all blast. I started the all-vs-all blast around 3:30pm and it's 7:10pm now and it's still running...</p>
<h2>Back to phylogenetic regression</h2>
<p>I've found a decent link here:</p>
<p><a href="http://www.mpcm-evolution.org/OPM/Chapter5_OPM/OPM_chap5.pdf">http://www.mpcm-evolution.org/OPM/Chapter5_OPM/OPM_chap5.pdf</a></p>
<p>that I've been working off of. I have the phylosignal package working and I think some of the results are similar? At least in terms of showing that there's some phylogenetic signal there. But I want a prediction of which genes are responsible.</p>
<p>So I've gone back into python and written a script called analyze_pfams.py. This goes into the generated pfam files for the S_epi strains for which data exists and scrapes them for interesting pfams to look at. The details of the way in which the pfams are picked can be changed pretty easily in the script, right now I have it in the following way:</p>
<p>First go through all the pfd files for strains involved in flow data and get every pfam that's in there. Every time I see a pfam in one of the files I count it. I then normalize so that for every strain I have a normalized count for the frequency of appearance for a given pfam term in that species's genome.</p>
<p>Next, I save the normalized frequencies for all the strains for each pfam. I calculate the mean frequency and standard deviation, as well as which strains the pfam_id shows up in.</p>
<p>Finally, to pick which pfams to move foward with, I save the pfam_ids with the following conditions:</p>
<pre><code>        if num_positive == 0 and num_negative == 3:
            to_append = True
        if num_negative == 0:
            to_append = True
        if abs(mean_negative - mean_positive) &gt; (pfam_std * 10):
            to_append = True
</code></pre>
<p>Note that as far as I can tell, the dominating condition here is the last one (pfam_std * 10). It might be worth quantifying this in a bit.</p>
<p>I write these all out as a .csv file (named 'interesting_pfams.csv'). I've imported this data back into R so that for each of the six strains, there is an associated frequency (I'm ignoring std for now) for each pfam_id. I have the associated counts as well. I've merged this data into a unified data.frame called df_merged (after wrestling around with the row names). And this plugs pretty nicely into the workflow here:</p>
<p><a href="http://www.mpcm-evolution.org/OPM/Chapter5_OPM/OPM_chap5.pdf">http://www.mpcm-evolution.org/OPM/Chapter5_OPM/OPM_chap5.pdf</a></p>
<p>I can get the individual plots of a single pfam_id vs CD8B_IL17A counts by using the following command:</p>
<pre><code>with(df_merged, plot(TetR_C_8 ~ mean, xlab = "TetR_C_8", ylab = "mean counts", main = "TetR_C_8 vs CD8B_IL17A Activation"))
</code></pre>
<p>Where in this case I'm plotting TetR_C_8 (which for the record looks decent?) But it looks like the p-value isn't all that high.</p>
<p>Well...while trying to sort that out, R just crashed. Gotta do a better job of saving workspaces as I go along. I'm trying to get back to the point where I was. I'm just running into an issue mergning again.</p>
<p>After a lot of finagling I've gotten a linear fit with brownian phylogenetic correction for the top (~150?) pfams that were chosen with the python script. Gonna paste the important commands here but also save a copy of the workspace and history into this post.</p>
<pre><code>residuals_mat &lt;- matrix(NA, ncol = (dim(dat)[2]), nrow=1)
colnames(residuals_mat) &lt;- colnames(dat)
residuals_mat
residuals_mat &lt;- residuals_mat[ , -1]
residuals_mat

for (i in 2:(dim(dat)[2])) {
    form &lt;- formula(paste(names(dat)[i], " ~ mean"))
    this.fit &lt;- gls(form, data=dat, correlation=corBrownian(phy=tree))
    residuals_mat[i - 1] &lt;- attributes(residuals(this.fit))$std[1]
}
</code></pre>
<p>Anyways...I'm getting too tired to write out all the commands specifically. But most of the source was adapted from the link above. The variables are saved to the workspace. Important things there are <strong>tree</strong> (loaded from tree file), <strong>dat</strong> = <strong>df_merged</strong> (merged data), <strong>df_residuals</strong> (calculated residuals for pfam ids) <strong>tree_final</strong> (phylo4d merged data with summary statistics of count data for use in phylosignal package). I also wrote the residuals file to a .csv so I have them seperately. By residuals I mean Residual standard error for example in this summary:</p>
<pre><code>Generalized least squares fit by REML
  Model: Phage_H_T_join ~ mean 
  Data: dat 
  Log-restricted-likelihood: 49.05774

Coefficients:
  (Intercept)          mean 
-8.143699e-05  8.619997e-04 

Correlation Structure: corBrownian
 Formula: ~1 
 Parameter estimate(s):
numeric(0)
Degrees of freedom: 6 total; 4 residual
Residual standard error: 0.0002399488 
</code></pre>
<p>Now that I'm plotting the lowest residual pfams, it looks like that isn't the metric I want to use. The low residual pfams (at least the ones I'm looking at) have all the points at zero counts (so that the counts are independent of the pfam in question. But the residuals are low because the best-fit line is just a straight line through the data. I need to look at another metric...maybe the p-value from the anova summary?</p>
<p>So I went through a similar process, now a couple of new variables:
<strong>df_pvals_sorted_1</strong> and <strong>df_pvals_sorted_2</strong> are the pfam ids sorted either by mean_pval or intercept_pval. It looks like there are two terms that are sort of at the top of both lists? One is the denaturase...and the other is DEDD_Tnp_IS110</p>

</div>






<pre style="display:none;" class="markdown"></pre>

<script>
  function renderRightAsMarkdown() {
      var right = document.getElementsByClassName("right");
      for (var i=0; i<right.length; i++) {
          src = right[i].innerHTML
          right[i].innerHTML = marked(src);
      }
  }
  renderRightAsMarkdown();
</script>




<script>

// Disable clipboard stuff for now
// var pre = document.getElementsByTagName('pre');
// for (var i = 0; i < pre.length; i++) {
//   var button = document.createElement('button');
//   button.className = 'btn fa fa-copy';
//   pre[i].appendChild(button);
// }

// var clipboard = new Clipboard('.btn', {
//   target: function(trigger) {return trigger.previousElementSibling;}
// });

// function showSuccess(elem){elem.setAttribute('class','btn fa fa-check');}
// function sleep (time) {return new Promise((resolve) => setTimeout(resolve, time));}
// clipboard.on('success',function(e){
//   showSuccess(e.trigger);
//   sleep(350).then(() => {e.clearSelection();
//     e.trigger.setAttribute('class','btn fa fa-copy');})
// });

// Experimental direct editing

// var paragraphs = $$("#entry>p");
// var lastEl = paragraphs.getLast();
// var path = '/projects/S_epi/2017-05-29';

// paragraphs.each(function(el) {
//   var dropArea = new Element('div', {'class':'right droparea',
//                                      'ondrop':'drop(event, path)',
//                                      'ondragover':'dragover_handler(event)'
//                                      });
//   var next = el.getNext();
//   if (!!next && !next.hasClass('right')) {
//     droptext = new Element('div', {'id':'targetString',
//                                    styles:{display: 'none'},
//                                    'html':el.get('html')
//                                   });
//     droptext.inject(dropArea);
//     dropArea.inject(el, 'after');
//   }
// });

// var rightimgs = $$("img.thumb");
// rightimgs.each(function(el) {
//   var button = document.createElement('button');
//   button.className = 'btn fa fa-window-close';
//   el.appendChild(button);
// });

// var markDownEl = document.querySelector(".markdown");

// new MediumEditor($$("#entry>p"), {
//     extensions: {markdown: new MeMarkdown(function (md) {markDownEl.textContent = md;})},
//     toolbar: {
//       allowMultiParagraphSelection: false,
//       buttons: ['bold', 'italic', 'underline']
//     },
//     disableReturn: true,
//     disableDoubleReturn: true,
//     disableExtraSpaces: true,
//     paste: {
//         forcePlainText: true,
//         cleanPastedHTML: true
//     }
// });

</script>

<a href="slides" id="play-slides-link"><i class="fa fa-play" aria-hidden="true"></i></a>


  </div>

  <footer>
    &copy; 2016. Made with <a target="_new" href="http://bionano.ucsf.edu/"><i class="fa fa-gears"></i></a> in SF.
  </footer>

</body>
